### PR TITLE
Source Hubspot: implement new stream to read associations in incremental mode

### PR DESCRIPTION
**Pull request by davydov-d**
## What
https://github.com/airbytehq/oncall/issues/359

## How
Implement new stream to read associations from a specialized API since API of the parent stream (used in incremental mode) does not support associations

### CODE REVIEW

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py [hunk 1]
```diff
@@ -807,6 +855,24 @@ def _process_search(
 
         return list(stream_records.values()), raw_response
 
+    def _read_associations(self, records: Iterable, identifiers: Iterable[Union[int, str]]) -> Iterable[Mapping[str, Any]]:
+        records_by_pk = {record[self.primary_key]: record for record in records}
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:859]: Why do we pass both identifiers and records if we can get identifiers from records?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py [hunk 2]
```diff
@@ -807,6 +855,24 @@ def _process_search(
...
+        identifiers = list(identifiers)
+        associations_stream = AssociationsStream(
+            api=self._api, start_date=self._start_date, credentials=self._credentials, parent_stream=self, identifiers=identifiers
+        )
+        slices = associations_stream.stream_slices(sync_mode=SyncMode.full_refresh)
+
+        for _slice in slices:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:866]: Do we need to update the schema of the output objects?

**Review by sherifnada — APPROVED** [review]
Just a question about schema but overall looks good

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py (hunk 2 above)
Comment (girarda) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:866]: I don't think this should change the schema as this is only a bug fix to retrieve the associations in incremental mode.
I looked at the `deals.json` stream as an example and it already has the `companies`, `contacts`, and `line_items` fields

**Review by girarda — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py (hunk 2 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:866]: Alexandre is right. Schemas already contain those fields, only data was missing in incremental mode

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py (hunk 1 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:859]: good point! fixed, thanks

**Pull request merged**